### ISSUE TITLE
Potentially set popups to `role="dialog"`

### ISSUE THREAD

**Issue by Malvoz** [Leaflet/Leaflet#7968]
**Is your feature request related to a problem? Please describe.**

Popups currently do not have any [semantics](https://developer.mozilla.org/en-US/docs/Glossary/Semantics).

**Describe the solution you'd like**

Set `leaflet-popup` to [`role="dialog"`](https://www.w3.org/TR/wai-aria/#dialog).

**Describe alternatives you've considered**

The [proposed HTML `<popup>` element](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/Popup/explainer.md), perhaps to replace `<div>` in the future, but not relevant now.

**Additional context**

Feedback from an accessibility expert is that the `dialog` role seems suitable for Leaflet's popups: https://github.com/w3c/aria/issues/1421#issuecomment-824152117.

**Comment by Malvoz** [comment]
For consideration:

Are there any conditions in which popups shouldn't be treated as dialogs?

The [`dialog`](https://www.w3.org/TR/wai-aria-1.1/#dialog) role doesn't have any hard requirements (in the WAI-ARIA 1.1 spec), though the role description states:

> Authors _**SHOULD**_ ensure that all dialogs (both modal and non-modal) have at least one focusable descendant element.
  
(note that "should" is not a "must")

Even if _all_ of the following popup options were set to `false`:

- [`closeButton`](https://leafletjs.com/SlavaUkraini/reference.html#popup-closebutton)
  (the default of this option fulfills the _recommendation_ "_**SHOULD**_ have at least one focusable descendant element")
- [`autoClose`](https://leafletjs.com/SlavaUkraini/reference.html#popup-autoclose)
  (the ability to close the element reinforces the idea that this is a (interactive) dialog)
- [`closeOnEscapeKey`](https://leafletjs.com/SlavaUkraini/reference.html#popup-closeonescapekey)
  (the default of this option fulfills the _expectation_ that (HTML) `<dialog>`s close on <kbd>Esc</kbd> (at least in some browsers))
- [`closeOnClick`](https://leafletjs.com/SlavaUkraini/reference.html#popup-closeonclick)
  (the ability to close the element reinforces the idea that this is a (interactive) dialog)

activating markers will still open/close the popups. As such I think `role="dialog"` is acceptable in any case, because the scenario with having all of the above options disabled _in a standalone popup_ is very unlikely? And even if that's the case, `dialog` is not a barrier to screen reader users.

To summarize: I believe that `role="dialog"` is appropriate, and I don't think it's strictly necessary to set the role conditionally.

**Comment by timveld** [comment]
@malvoz I agree that the proposed dialog role is (formally speaking) in accordance with the wai-aria specification.

However the question is whether it is an appropriate pattern for a popup with a single close button.

A standalone popup is simply a piece of information the author wants me to read and may allow me to hide using a close button. It is not supposed to interrupt my workflow and does not contain interactive elements. Therefore, it is not (semantically) a dialog. The problem that the popup doesn't have sufficient context for non-visual users is better addressed by placing a heading above it (responsibility of the Leaflet implementor). Alternatively we could consider a region landmark which makes clear that it is an information popup.

Popups which appear in response to the end user clicking a marker *can* be considered dialogs, because they interrupt the workflow (so the user first reads the message, then dismisses the popup and returns to the workflow). In that case proper focus management is essential to prevent confusion; when the popup opens focus should move to the dialog (see issue #2199) and when the popup closes focus must be set back on the associated marker.

However, designing such an interaction well is very complicated. Besides the many focus management and trapping issues, the way different screen readers on different platforms handle dialogs is far from uniform and I see a lot of potential for screen reader users to get lost / confused. Long story short, we are giving ourselves and screen reader users a lot of headaches by doing this and I think it is far more complicated than it needs to be.


A far simpler approach is to follow the accordion pattern: we consider markers to be things which expand when you click them and collapse when you click them again. Visually it looks a bit different, but to a screen reader user it will feel like a perfectly normal accordion with the addition of close buttons which are just shortcuts / equivalents to collapsing the marker.


The main steps we should take to implement this:
* Place aria-expanded on the markers (issue #7527)
* Ensure that the popup appears immediately below  the marker in the DOM

* Make sure that when the close button is clicked, focus is returned to the marker which opened it
* Pressing escape should be equivalent to clicking the close button (there could be an argument that it is confusing for blind users, but in my opinion this is not going to be an issue in practice and won't cause WCAG violations)
* We can also close this dreadful issue #2199 because placing focus on the close button is no longer necessary, tab order will be logical given the DOM ordering


What do readers think of this solution?

**Comment by Malvoz** [comment]
@timveld Thanks for the feedback.

> A standalone popup is simply a piece of information the author wants me to read and may allow me to hide using a close button. It is not supposed to interrupt my workflow and does not contain interactive elements. Therefore, it is not (semantically) a dialog.

Are you thinking of _modal_ dialogs ([`aria-modal="true"`](https://www.w3.org/TR/wai-aria-1.1/#aria-modal))? Modal dialogs usually trap focus and hide the underlying content whilst the modal dialog is open. The definition of `role="dialog"` states:

> A dialog that is designed to interrupt workflow is usually modal.

> authors **_SHOULD_** manage focus of modal dialogs

`role="dialog"` is non-modal by default, I am not proposing modal dialogs.

> The problem that the popup doesn't have sufficient context for non-visual users is better addressed by placing a heading above it (responsibility of the Leaflet implementor).

I think the problems with such an approach that expects headings are: 1) many developers don't use headings in their popups, 2) even if they did, that would only immediately accommodate AT users who happen to [navigate by headings](https://dequeuniversity.com/screenreaders/nvda-keyboard-shortcuts#:~:text=Headings:) or read the document from top to bottom? A user who would e.g. <kbd>tab</kbd> to the Close button wouldn't have any context for it.

If I set `leaflet-popup` to `role="dialog"` and <kbd>tab</kbd> to the Close button, NVDA announces "dialog, button, close popup" which gives the button (or any other content provided by the developer) context, by default.

> Alternatively we could consider a region landmark which makes clear that it is an information popup.

I don't think `region` is appropriate, as [landmarks should be used sparingly](https://markwilcock.com/2020/07/accessibility-rich-internet-applications-aria-landmarks/):

> The landmark role is intended to be used sparingly, landmarks should only be used to outline the overall segments of the page. If too many landmark roles are applied to a page, it can make the advantages gain by applying these roles obsolete or even cause a detrimental effect to users.

While I generally agree with the rest of your response, it is off-topic for this issue. Issue https://github.com/Leaflet/Leaflet/issues/2199 is only one aspect of accessible marker/popup navigation. Perhaps we should open a new issue which supersedes that one, in which we can discuss the requirements for a logical sequential navigation and focus management between markers and popups.
